# Trilha A: Plataforma Multicloud com IA

Vamos iniciar o projeto coletando informações da API da AWS.

In [5]:
import boto3
import pandas as pd
import json
from itertools import product
from time import sleep

Vamos coletar algumas configurações de maquinas que consistem em várias combinações de CPU, memória, armazenamento, localização das máquinas, sistema operacional e outras opções. O objetivo é obter uma variedade de opções e entender quais delas é melhor para uma atividade. 
Iniciamos com uma categorização que a própria a AWS fornece em seu site, são elas:

- Uso geral: Fornecendo um equilíbrio de recursos de computação, memória e rede.
- Computação Acelerada: Usam  aceleradores de hardware, ou coprocessadores, para executar funções, como cálculo de número de pontos flutuante, processamento de gráficos e entre outras atividade.
- Otimizadas para armazenamento: Aqui temos instâncias projetadas para cargas de trabalho que exigem acesso de leitura e gravação sequencial alto a conjuntos de dados muito grandes no armazenamento local.


In [22]:
# Criando cliente

client = boto3.client('pricing', region_name='us-east-1', aws_access_key_id='AKIAVUQSHW3BUOSFFM4O',
    aws_secret_access_key='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')

# Categorias de instâncias
instancias = {
    'Uso Geral': {
        't3.nano':      {'vCPU': 2,  'RAM': 0.5,   'GPU': 0},
        't3.micro':     {'vCPU': 2,  'RAM': 1,     'GPU': 0},
        't3a.small':    {'vCPU': 2,  'RAM': 2,     'GPU': 0},
        't3.medium':    {'vCPU': 2,  'RAM': 4,     'GPU': 0},
        'm5.large':     {'vCPU': 2,  'RAM': 8,     'GPU': 0},
        'm5.xlarge':    {'vCPU': 4,  'RAM': 16,    'GPU': 0},
        'm6g.medium':   {'vCPU': 1,  'RAM': 4,     'GPU': 0}
    },
    'Memória': {
        'r8g.medium':  {'vCPU': 1,  'RAM': 8,    'GPU': 0},
        'r8g.large':   {'vCPU': 2,  'RAM': 16,   'GPU': 0},
        'r8g.xlarge':  {'vCPU': 4,  'RAM': 32,   'GPU': 0},
        'r8gd.medium': {'vCPU': 1,  'RAM': 8,    'GPU': 0},
        'r8gd.large':  {'vCPU': 2,  'RAM': 16,   'GPU': 0},

    },
    'Computação Acelerada': {
        'g4dn.xlarge':   {'vCPU': 4,  'RAM': 16,    'GPU': 1},
        'g5.xlarge':     {'vCPU': 4,  'RAM': 16,    'GPU': 1},
        'p2.xlarge':     {'vCPU': 4,  'RAM': 61,    'GPU': 1},
        'p3.2xlarge':    {'vCPU': 8,  'RAM': 61,    'GPU': 1},
        'p4d.24xlarge':  {'vCPU': 96, 'RAM': 1152,  'GPU': 8}
    },
    'Armazenamento': {
        'i3.large':      {'vCPU': 2,  'RAM': 15.25, 'GPU': 0},
        'i3en.large':    {'vCPU': 2,  'RAM': 16,    'GPU': 0},
        'd2.xlarge':     {'vCPU': 4,  'RAM': 30.5,  'GPU': 0},
        'd3.xlarge':     {'vCPU': 4,  'RAM': 32,    'GPU': 0},
        'h1.2xlarge':    {'vCPU': 8,  'RAM': 32,    'GPU': 0}
    }
}

regioes = ['US East (N. Virginia)', 'US West (Oregon)', 'Europe (Ireland)', 'South America (Sao Paulo)', 'Asia Pacific (Tokyo)']

sistema = ['Linux', 'Windows', 'RHEL', 'SUSE']

tenancy = ['Shared', 'Dedicated']

compra = ['On-Demand', 'Reserved']

# Dados coletados
resultados = []

for categoria, tipos in instancias.items():
    for tipo, specs in tipos.items():
        for regiao, so, ten in product(regioes, sistema, tenancy):
            try: 
                response = client.get_products(
                    ServiceCode='AmazonEC2',
                    Filters=[
                        {'Type': 'TERM_MATCH', 'Field':'location','Value':regiao},
                        {'Type': 'TERM_MATCH', 'Field': 'instanceType', 'Value': tipo},
                        {'Type': 'TERM_MATCH', 'Field': 'operatingSystem', 'Value': so},
                        {'Type': 'TERM_MATCH', 'Field': 'preInstalledSw', 'Value': 'NA'},
                        {'Type': 'TERM_MATCH', 'Field': 'tenancy', 'Value': ten},
                        {'Type': 'TERM_MATCH', 'Field': 'capacitystatus', 'Value': 'Used'},
                    ],
                    MaxResults=1
                )

                preco = None
                produto = json.loads(response['PriceList'][0])
                termos = produto['terms']['OnDemand']
                for term_key in termos:
                    preco = list(termos[term_key]['priceDimensions'].values())[0]['pricePerUnit']['USD']

                resultados.append({
                    'Categoria': categoria,
                    'Instância': tipo,
                    'Região': regiao,
                    'SO': so,
                    'Tenancy': ten,
                    'vCPU': specs['vCPU'],
                    'RAM_GiB': specs['RAM'],
                    'GPU': specs['GPU'],
                    'Preço_USD': float(preco) if preco else None
                })

            except Exception as e:
                print(f'Erro em {tipo} | {regiao} | {so} | {ten}: {e}')
                continue
            sleep(0.2)



# Converter para DataFrame
df = pd.DataFrame(resultados)


Erro em t3.nano | US East (N. Virginia) | RHEL | Shared: list index out of range
Erro em t3.nano | US East (N. Virginia) | RHEL | Dedicated: list index out of range
Erro em t3.nano | US West (Oregon) | RHEL | Shared: list index out of range
Erro em t3.nano | US West (Oregon) | RHEL | Dedicated: list index out of range
Erro em t3.nano | Europe (Ireland) | Linux | Shared: list index out of range
Erro em t3.nano | Europe (Ireland) | Linux | Dedicated: list index out of range
Erro em t3.nano | Europe (Ireland) | Windows | Shared: list index out of range
Erro em t3.nano | Europe (Ireland) | Windows | Dedicated: list index out of range
Erro em t3.nano | Europe (Ireland) | RHEL | Shared: list index out of range
Erro em t3.nano | Europe (Ireland) | RHEL | Dedicated: list index out of range
Erro em t3.nano | Europe (Ireland) | SUSE | Shared: list index out of range
Erro em t3.nano | Europe (Ireland) | SUSE | Dedicated: list index out of range
Erro em t3.nano | South America (Sao Paulo) | RHEL |

Podemos visualizar o nosso DataFrame com as informações obtidas. 

In [23]:
df.head()

Categoria Instância                 Região       SO    Tenancy  vCPU  \
0  Uso Geral   t3.nano  US East (N. Virginia)    Linux     Shared     2   
1  Uso Geral   t3.nano  US East (N. Virginia)    Linux  Dedicated     2   
2  Uso Geral   t3.nano  US East (N. Virginia)  Windows     Shared     2   
3  Uso Geral   t3.nano  US East (N. Virginia)  Windows  Dedicated     2   
4  Uso Geral   t3.nano  US East (N. Virginia)     SUSE     Shared     2   

   RAM_GiB  GPU  Preço_USD  
0      0.5    0     0.0052  
1      0.5    0     0.0052  
2      0.5    0     0.0052  
3      0.5    0     0.0098  
4      0.5    0     0.0052

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 546 entries, 0 to 545
Data columns (total 9 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Categoria  546 non-null    object 
 1   Instância  546 non-null    object 
 2   Região     546 non-null    object 
 3   SO         546 non-null    object 
 4   Tenancy    546 non-null    object 
 5   vCPU       546 non-null    int64  
 6   RAM_GiB    546 non-null    float64
 7   GPU        546 non-null    int64  
 8   Preço_USD  546 non-null    float64
dtypes: float64(2), int64(2), object(5)
memory usage: 38.5+ KB


Temos ao todo 546 tipos de máquinas localizadas em 5 regiões, com diferentes configurações. Os dados foram obtidos de forma consistentes, sem aparesentar dados faltantes e com a tipagem de acordo com as suas colunas. Nosso objetivo nesta fazer é fazer uma análise exploratória para identificar as características dessas máquinas e suas peculiaridades.

Para direcionar esta análise, iremos propor algumas perguntas para serem respondidas.

1. Quais regiões são mais caras?
2. Existe diferença de custo entre sistemas operacionais?
3. Qual instância oferece melhor custo-benefício?
4. O preço varia para a mesma instância em diferentes regiões?
5. Instâncias com GPU são mais caras?
6. Qual o custo médio por vCPU e por GiB de RAM?
7. Tenancy influência o preço?
8. Existe correlação entre recursos e preço?
9. Qual o perfil das instâncias mais caras/baratas?
10. Conseguimos criar um Clusters de instâncias com características similares?



## Quais regiões são mais caras?
Nosso objetivo é verificar variações médias de preço por região, para tal vamos analisar o